### PR TITLE
Update cache after clearing delimiters in CodeEdit

### LINKED ISSUE #49568: Executing `CodeEdit.get_delimiter_start_position` crashes Godot
### Godot version

4.0.dev.custom_build. e82a1113a

### System information

Ubuntu 21.04 - Nvidia GTX 970, Gnome shell 3.38 X11

### Issue description

When executing
```
func _process(_delta):
	var q = CodeEdit.new()
	q.add_comment_delimiter(".", ".", true)
	q.clear_string_delimiters()
	q.get_delimiter_start_postion(0, -5)
```

Godot crashes with backtrace
```
ERROR: FATAL: Index p_index = 0 is out of bounds (size() = 0).
   at: get (./core/templates/cowdata.h:157)
handle_crash: Program crashed with signal 4
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] godot4s() [0x2050562] (/mnt/Miecz/godot/platform/linuxbsd/crash_handler_linuxbsd.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x41040) [0x7fbd9fd3b040] (??:0)
[3] CowData<Map<int, int, Comparator<int>, DefaultAllocator> >::get(int) const (/mnt/Miecz/godot/./core/templates/cowdata.h:157 (discriminator 7))
[4] Vector<Map<int, int, Comparator<int>, DefaultAllocator> >::operator[](int) const (/mnt/Miecz/godot/./core/templates/vector.h:89)
[5] CodeEdit::get_delimiter_start_position(int, int) const (/mnt/Miecz/godot/scene/gui/code_edit.cpp:686 (discriminator 8))
[6] void call_with_variant_args_retc_helper<__UnexistingClass, Vector2, int, int, 0ul, 1ul>(__UnexistingClass*, Vector2 (__UnexistingClass::*)(int, int) const, Variant const**, Variant&, Callable::CallError&, IndexSequence<0ul, 1ul>) (/mnt/Miecz/godot/./core/variant/binder_common.h:686 (discriminator 4))
[7] void call_with_variant_args_retc_dv<__UnexistingClass, Vector2, int, int>(__UnexistingClass*, Vector2 (__UnexistingClass::*)(int, int) const, Variant const**, int, Variant&, Callable::CallError&, Vector<Variant> const&) (/mnt/Miecz/godot/./core/variant/binder_common.h:444)
[8] MethodBindTRC<Vector2, int, int>::call(Object*, Variant const**, int, Callable::CallError&) (/mnt/Miecz/godot/./core/object/method_bind.h:543)
[9] Object::call(StringName const&, Variant const**, int, Callable::CallError&) (/mnt/Miecz/godot/core/object/object.cpp:816 (discriminator 1))
[10] Variant::call(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (/mnt/Miecz/godot/core/variant/variant_call.cpp:981)
[11] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/mnt/Miecz/godot/modules/gdscript/gdscript_vm.cpp:1473)
[12] GDScriptInstance::call(StringName const&, Variant const**, int, Callable::CallError&) (/mnt/Miecz/godot/modules/gdscript/gdscript.cpp:1487)
[13] ScriptInstance::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/mnt/Miecz/godot/core/object/script_language.cpp:311)
[14] Node::_notification(int) (/mnt/Miecz/godot/scene/main/node.cpp:58)
[15] Node::_notificationv(int, bool) (/mnt/Miecz/godot/./scene/main/node.h:45 (discriminator 14))
[16] CanvasItem::_notificationv(int, bool) (/mnt/Miecz/godot/./scene/main/canvas_item.h:164 (discriminator 3))
[17] Node2D::_notificationv(int, bool) (/mnt/Miecz/godot/./scene/2d/node_2d.h:37 (discriminator 3))
[18] Object::notification(int, bool) (/mnt/Miecz/godot/core/object/object.cpp:827)
[19] SceneTree::_notify_group_pause(StringName const&, int) (/mnt/Miecz/godot/scene/main/scene_tree.cpp:809)
[20] SceneTree::process(float) (/mnt/Miecz/godot/scene/main/scene_tree.cpp:443 (discriminator 2))
[21] Main::iteration() (/mnt/Miecz/godot/main/main.cpp:2495)
[22] OS_LinuxBSD::run() (/mnt/Miecz/godot/platform/linuxbsd/os_linuxbsd.cpp:278)
[23] godot4s(main+0x3fc) [0x204e042] (/mnt/Miecz/godot/platform/linuxbsd/godot_linuxbsd.cpp:60)
[24] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xd5) [0x7fbd9fd22565] (??:0)
[25] godot4s(_start+0x2e) [0x204db8e] (??:?)
```

This example was found by fuzzer, so I don't recommend trying understand in what situation such code could be used in real project, because such situation probably doesn't exists.

### Steps to reproduce

Above

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by timothyqiu**
Fixes #49568.

It crashed because the cache is out of sync with the real world :(

### CODE REVIEW

File: scene/gui/code_edit.cpp [hunk 1]
```diff
@@ -1526,6 +1526,7 @@ void CodeEdit::_clear_delimiters(DelimiterType p_type) {
 		}
 	}
 	delimiter_cache.clear();
+	_update_delimiter_cache();
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1529]: Should only do this if `setting_delimiters == false` else we'll calculate it twice in short succession, which can be quite expensive.

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Nice find, minor performance tweak commented inline, but looks good otherwise.

File: scene/gui/code_edit.cpp [hunk 2]
```diff
@@ -1525,7 +1525,10 @@ void CodeEdit::_clear_delimiters(DelimiterType p_type) {
 			delimiters.remove(i);
 		}
 	}
-	delimiter_cache.clear();
+	if (!setting_delimiters) {
+		delimiter_cache.clear();
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1529]: Apologises, should have been more clear, still need to call `delimiter_cache.clear()`, only need to skip `_update_delimiter_cache` else the cache will grow quite large.

Comment (timothyqiu) [reply scene/gui/code_edit.cpp:1529]: That did come across my mind when I was doing the change, I should have asked 🤣

**Review by Paulb23 — APPROVED** [review]
Looks good.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**